src/snakemake/rules.py

```
@@ -1053,97 +1061,84 @@
+            except ResourceConstraintError:
+                global_val = global_res.value
+                global_type = (
+                    "an int" if isinstance(global_val, int) else type(global_val)
+                )
+                raise WorkflowError(
+                    f"Resource {resource} is of type {type(val.value).__name__} but "
+                    f"global resource constraint defines {global_type} with value "
+                    f"{global_val}. Resources with the same name need to have the same "
+                    "types (int, float, or str are allowed).",
+                    rule=self,
+                )
```

Ruff (0.11.9)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)